Add guide for retiring OpenStack releases (#994)
osism/osism.github.io: docs/guides/developer-guide/retiring-openstack-releases.md

| `openstack-octavia-amphora-image` | Drop the per-version build jobs from `.zuul.yaml`, the version rows in `README.md`, and the version block in `playbooks/cleanup.yml`.                                                                                                             | [#142](https://github.com/osism/openstack-octavia-amphora-image/pull/142) (2024.2) |
| `container-images-kolla`          | Remove the per-version build branch in `.zuul.yaml` and the matching `elif` in `scripts/001-prepare.sh`.                                                                                                                                          | [#573](https://github.com/osism/container-images-kolla/pull/573) (2023.2)          |
| `container-image-kolla-ansible`   | Remove the per-version build/push jobs in `.zuul.yaml` and the matching `elif` in the `Containerfile`.                                                                                                                                            | [#745](https://github.com/osism/container-image-kolla-ansible/pull/745) (2023.2)   |

> | `metalbox`                        | Remove `zuul/vars/container-images-openstack-<version>.yml`, the per-image entries in `zuul/vars/container-images-metalbox.yml`, the version from the `openstack_versions` default in `zuul/mirror-octavia-image.yml`, and any `.zuul.yaml` jobs. | (no prior precedent — see note)                                                    |

Reviewer: testbed repository missing (cleanup of zuul configuration)

> | `metalbox`                        | Remove `zuul/vars/container-images-openstack-<version>.yml`, the per-image entries in `zuul/vars/container-images-metalbox.yml`, the version from the `openstack_versions` default in `zuul/mirror-octavia-image.yml`, and any `.zuul.yaml` jobs. | (no prior precedent — see note)                                                    |

Reviewer: openstackclient image build in container-images should also be disabled

---
sidebar_label: Retiring OpenStack Releases
sidebar_position: 15
---

# Retiring OpenStack Releases

## Policy

We retire an OpenStack release when it is retired upstream — the version is
removed from every repository that references it. When sites still run an

> end-of-life (EOL) release and we owe them security updates, we **delay** the

Reviewer 2: ```suggestion
removed from every repository that references it. In case of a known
interest of an operator to continue running this release, we may **delay** the
```

removal and keep building patched images from the `<version>-eol` tag in the
meantime (see [Retiring a release](#retiring)). Retirement still happens; it is
only postponed until security support ends.

When a release is retired upstream depends on whether it is a SLURP release.
The authoritative schedule is the [OpenStack
releases](https://releases.openstack.org/) page; do not hardcode EOL dates,

> look them up there.

Reviewer 2: Who does the "do not hardcode" address here? Where would this someone hardcode them?

Author: Removed.

## Upstream release lifecycle

Every code change in this guide is driven by the way upstream OpenStack moves a

> release through its lifecycle:

Reviewer 2: ```suggestion
Every code change described here is driven by the way upstream OpenStack moves a
release through its lifecycle:
```

Reviewer 2: I stumbled here a bit here, since i thought "Why would code changes to this guide align with anything upstream, this is a developer guide for OSISM workflows". So I hope these suggestions makes it more obvious what is meant here.

   example `stable/2024.2`). This is what most build jobs check out.
2. **Unmaintained** — the branch is renamed to `unmaintained/<version>` (for
   example `unmaintained/2024.1`). It still exists, but upstream no longer
   guarantees maintenance.
3. **End of life (EOL)** — the branch is **deleted**. Its final state is
   preserved as the immutable `<version>-eol` tag (for example `2024.2-eol`).

Anything that referred to `stable/<version>` or `unmaintained/<version>`
**breaks** the moment upstream deletes that branch. That broken build is
usually the first visible sign that a release has reached EOL.

### SLURP vs. non-SLURP

SLURP (Skip-Level Upgrade Release Process) releases are maintained for longer
than the non-SLURP releases in between them. As a result two adjacent releases
can be in different phases at the same time. For example:

- `2024.1` (SLURP) — still alive, in the `unmaintained/2024.1` phase.
- `2024.2` (non-SLURP) — already EOL; `stable/2024.2` is gone, only the
  `2024.2-eol` tag remains.

This is why a non-SLURP release is typically the first one you need to retire,
while the older SLURP release next to it is still being built.

> ## Retiring a release {#retiring}

Reviewer 2: what does the {#retiring} do here?

Author: Markdown code for creating an anchor.

Retirement always ends the same way: the version is removed from every
repository that references it. The only open question is *when*. Work through
this once the release reaches EOL upstream:

1. **The trigger.** Upstream deletes `stable/<version>`, so every build that
   checked it out fails. That broken build is usually the first sign the
   release has reached EOL.

> 2. **Do any sites still run this version that we owe security updates?**

Reviewer 2: ```suggestion
2. **Do we know any operators running this version that still need security updates?**
```

   - **No** — remove the version now. Go straight to [Removing the
     version](#removing).
   - **Yes** — we cannot remove it yet, because we still need to ship security
     updates. First [keep the affected images building from the `<version>-eol`
     tag](#keep-building) as a stopgap, then defer the removal until security

>      support ends.

Reviewer 2: ```suggestion
     support ends or all operators have upgraded to a maintained release version.
```

  vehicle for a fix, so they keep building from the `<version>-eol` tag.
- `cfg-cookiecutter` only scaffolds a *new* configuration repository. An existing
  site generated its configuration once and never re-runs cookiecutter for a
  patch, so it is removed now.
- `openstack-octavia-amphora-image` builds the Octavia amphora appliance image.
  Removing it means no fresh amphora builds for the EOL release — a deliberate
  scope decision that amphora-level rebuilds are not part of the EOL
  security-update commitment.

The rule of thumb: keep the repositories that carry patched OpenStack code to a
running site; remove the ones only needed to *stand up* a new deployment, which
should not happen after EOL.

:::note Reviewer check
The split above reflects how 2024.2 was handled, where
`container-images-kolla` and `container-image-kolla-ansible` received the
`-eol` build fix while `cfg-cookiecutter` and `openstack-octavia-amphora-image`
were removed. If a future release needs amphora security updates for a paying
site, `openstack-octavia-amphora-image` would have to keep building too —
re-evaluate the list against the actual support commitment for the release

> being retired.

Reviewer 2: ```suggestion
were removed. If a future release needs amphora security updates for a
site, `openstack-octavia-amphora-image` would have to keep building too —
re-evaluate the list against the actual support commitment for the release
being retired.
```

fi
```

The same pattern applies to any other build that checks out a per-version
branch: point the retired version at its `<version>-eol` tag instead of
`stable/<version>`.

The intermediate `unmaintained/<version>` phase (the `2024.1` line above) is
handled the same way — point at `origin/unmaintained/<version>` while the
branch still exists, then switch to the `<version>-eol` tag once it is deleted.

### Defer the removal

Keeping the images building is a stopgap, not the end state. Prepare the
removal changes (see [Removing the version](#removing)) for these build
repositories anyway, open them as pull requests, and **hold them open** until
we no longer ship security updates for the release. When that window closes,
merge the held PRs to complete the retirement. Keeping the removal PRs open
(rather than only filing an issue) makes the pending work visible and ready to
merge.

For example, while 2024.2 stays on the `-eol` tag, the removal PRs
[container-images-kolla #726](https://github.com/osism/container-images-kolla/pull/726)
and
[container-image-kolla-ansible #908](https://github.com/osism/container-image-kolla-ansible/pull/908)

> remain open and are merged once 2024.2 security support ends.

Reviewer 2: ```suggestion
remain open and are merged once 2024.2 can be fully retired and removed.
```

## Removing the version {#removing}

This is the end state of every retirement, reached either immediately or after
the security-update bridge above. Remove the version from every repository that
carries a per-version reference to it. The repositories below encode
per-OpenStack-version build matrices; use the linked changes as worked

> examples.

Reviewer 2: ```suggestion
per-OpenStack-version build matrices; use the linked changes as examples.
```
